### ISSUE TITLE
Custom LSP Integration

### ISSUE THREAD

**Issue by zbirenbaum** [zbirenbaum/copilot.lua#209]
So my custom copilot-lsp implementation https://github.com/zbirenbaum/copilot-rs finally seems to be in a pretty decent state and can now be used as a drop in replacement excluding panel completions, which means its time to start considering integration here. This is mostly an RFC post regarding a couple of things:

### How should we handle distribution of the alternative LSP? 
We could distribute an equivalent `agent.js` using wasm bindings but there two points for consideration:
- Since copilot-rs does not implement telemetry it is very possibly going to get shutdown
- We should avoid any backlash on copilot.lua if copilot-rs gets shut down

### How should we handle feature integration?
The goal of developing an alternative LSP client has always been to achieve (at least) the following:
- Allow for VSCode only features such as copilot chat, which are trivial to implement on the LSP side and just involve sending off a custom prompt
- Allow for increased customization of parameters exposed by the actual github copilot api endpoint (it's literally an OpenAI endpoint, just check their docs) such as number of completions, tokens, etc...
- Allow for integration of other models such as LLAMA by setting up middleware #38
Since these features will eventually be present in copilot-rs, but not the official agent.js, client side support becomes a bit more complicated
- Make support for features like workspaces clear and documented #171  

@MunifTanjim would love to hear your thoughts. I'm leaning towards creating a new fork of copilot.lua just to support and redistribute copilot-rs and putting a note about it in the README

**Comment by MunifTanjim** [comment]
Even if it's included in this repo, it should be somewhat separated (either opt-in or opt-out) from the current implementation that mimics everything from https://github.com/github/copilot.vim. Then the user can choose which implementation to use.

**Comment by zbirenbaum** [comment]
> Even if it's included in this repo, it should be somewhat separated (either opt-in or opt-out) from the current implementation that mimics everything from https://github.com/github/copilot.vim. Then the user can choose which implementation to use.

Yeah this was what I was thinking. I'm really hesitant to put it in this repo at all unless I can find some precedent to show it's remotely safe. It doesn't break any license terms but not sure that will matter.

**Comment by caniko** [comment]
@zbirenbaum what is missing in `copilot-rs`? Just testing?

Also, I noticed there is no documentation; why?

**Comment by S1M0N38** [comment]
Official copilot LSP:
https://www.npmjs.com/package/@github/copilot-language-server

**Comment by AntoineGS** [comment]
I am guessing this is not on the horizon anymore :) Closing for the sake of reducing open Issues but feel free to reopen if you are still planning to do this!
